depends: Allow building with system clang

dongcarl opened the pull request:
This replaces: #17099

-----

This patchset allows us to force depends to use system clang.

Previously, #17099 removes our dependency on a specific clang we download from llvm.org, but theuni pointed out that since OSX builds are only ever built with a version of clang that is chosen and "blessed" by Apple, it is more likely that the user will encounter problems if they use their system clang. This patchset forces the user to set `FORCE_USE_SYSTEM_CLANG=1` in order to use their system clang (when they know what they're doing)

dongcarl committed:
depends: Decouple toolchain + binutils

For now they remain the same, but in the next commit, we will assign
them differently according to wether or not we're using system clang.

dongcarl committed:
depends: Allow building with system clang

MarcoFalke commented:
wen rebase, sir?

MarcoFalke commented:
To give some context, I am still running into issues cross-compiling for mac on non-x86 arch. So I am hoping that this pull improves it for me.


```
Preprocessing native_cctools...
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
-- The ASM compiler identification is unknown
-- Found assembler: /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang
-- Check for working C compiler: /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang
-- Check for working C compiler: /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang -- broken
CMake Error at /usr/share/cmake-3.10/Modules/CMakeTestCCompiler.cmake:52 (message):
  The C compiler

    "/home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/libtapi/build/CMakeFiles/CMakeTmp
    
    Run Build Command:"/usr/bin/make" "cmTC_09e5c/fast"
    make[1]: Entering directory '/home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/libtapi/build/CMakeFiles/CMakeTmp'
    /usr/bin/make -f CMakeFiles/cmTC_09e5c.dir/build.make CMakeFiles/cmTC_09e5c.dir/build
    make[2]: Entering directory '/home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/libtapi/build/CMakeFiles/CMakeTmp'
    Building C object CMakeFiles/cmTC_09e5c.dir/testCCompiler.c.o
    /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang    -o CMakeFiles/cmTC_09e5c.dir/testCCompiler.c.o   -c /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/libtapi/build/CMakeFiles/CMakeTmp/testCCompiler.c
    /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang: 1: /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang: ELF: not found
    /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang: 2: /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang: @!: not found
    /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang: 7: /home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/toolchain/bin/clang: Syntax error: ")" unexpected
    CMakeFiles/cmTC_09e5c.dir/build.make:65: recipe for target 'CMakeFiles/cmTC_09e5c.dir/testCCompiler.c.o' failed
    make[2]: *** [CMakeFiles/cmTC_09e5c.dir/testCCompiler.c.o] Error 2
    make[2]: Leaving directory '/home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/libtapi/build/CMakeFiles/CMakeTmp'
    Makefile:126: recipe for target 'cmTC_09e5c/fast' failed
    make[1]: *** [cmTC_09e5c/fast] Error 2
    make[1]: Leaving directory '/home/travis/build/MarcoFalke/bitcoin-core/depends/work/build/x86_64-apple-darwin16/native_cctools/3764b223c011574971ee3ae09ce968ba5dc2f00f-fb072d1ddcc/libtapi/build/CMakeFiles/CMakeTmp'

dongcarl commented:
Rebased over #18072 and updated PR description.

dongcarl commented:
@MarcoFalke This will definitely improve the situation for you (as I think the error you encounter is because we download a clang from llvm.org that's meant to be run on an x86_64 machine), but not sure if it will solve _everything_.

MarcoFalke commented:
Looks like travis may be unable to compile mac depends, but the depends stderr is not shown. See https://github.com/bitcoin/bitcoin/issues/16368#issuecomment-638219853

dongcarl commented:
> Looks like travis may be unable to compile mac depends, but the depends stderr is not shown. See [#16368 (comment)](https://github.com/bitcoin/bitcoin/issues/16368#issuecomment-638219853)

I think it's cuz we need the regenerated macOS SDK with the libc++ headers

theuni committed:
depends: force a new host id string if FORCE_USE_SYSTEM_CLANG is in use

This should be caught by the differing clang --version outputs, but because we
haven't yet extracted our pinned clang, the system one is actually used for the
version check. That's not a problem because bumping our pinned clang will
cause a rebuild of everything anyway.

theuni commented:
Concept ACK and this looks to be in good shape once rebased on top of feedback/fixups in #19240.

The changes in native_cctools.mk are pretty ugly but straightforward, and current behavior is unchanged as long as FORCE_USE_SYSTEM_CLANG remains unused. Though I think @dongcarl has finally convinced me that we'll probably move on to system-clang by default at some point and drop our weird pinned one. So hopefully the ugliness is only temporary.

Well done!

theuni commented:
Travis should be able to find MacOSX10.14-with-libcxx-headers.sdk.tar.gz now :)

fanquake commented:
Given that we are going to be using a system Clang, I had a bit of a look into what patching Ubuntu and Guix are doing with their Clang packages to see if there was anything we needed to consider; i.e similar to #18921.

#### Guix

The Clang and LLVM patches are [here](https://git.savannah.gnu.org/cgit/guix.git/tree/gnu/packages/patches). They seem fairly mundane. Mostly adjusting include/search paths, and fixing sanitizer build & ABI related issues.

#### Ubuntu/Debian
For Clang 6.0, the most macOS related patch seems to be [silent-test-macho.diff](https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/blob/8/debian/patches/silent-test-macho.diff). Looks like it deletes one of the tests for `llvm-objdump` and macho files. Added in this [commit](https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/commit/3abc9f813c4cc285ced926c37a4c592d939c199c). The Clang package doesn't seem to be modified like GCC, where certain hardening options are defaulted to ON etc. 

For Clang-10, one additional macOS related [patch](https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/blob/10/debian/patches/remove-apple-clang-manpage.diff) has landed, which removes macOS specific info, i.e `-mmacosx-version-min`, from the man page. The options still exist in Clang, but seem to have been removed from the manpage just to make it less "Mac-centric". Added in this [commit](https://salsa.debian.org/pkg-llvm-team/llvm-toolchain/-/commit/3232ed4c0f8c594043d3b3ee49d88527f228030c). Related bug report: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=743133.

dongcarl commented:
PR rebased, Gitian build succeeded:
```yml
- out_manifest: |
    7e218500ca12ba66d808041c2bfb46bcc2edfefa9a7d4518bf920f6665e39d73  bitcoin-842e81503b3b-osx-unsigned.dmg
    4619e5b8cbf9d09c02a4d10fb1ac0c8d36acb59d42f32ded743f604e66ac9072  bitcoin-842e81503b3b-osx-unsigned.tar.gz
    e1f8727102431a6a441b05d14c6fcc3e5b4d7017e709c3744379f8c0187a6e61  bitcoin-842e81503b3b-osx64.tar.gz
    f779ba3b880b0a79e6defe5de72dd82a27a3440f13e8144e1849b1e52bd3476a  src/bitcoin-842e81503b3b.tar.gz
```

dongcarl commented:
PR rebased again, Gitian build results:
```yml
- out_manifest: |
    98649e59d846ffcd56837b9128a9c4c5bd98684f4c8b6051ebf1f2f4dded4c17  bitcoin-03a7aaab8409-osx-unsigned.dmg
    ecbc41ac5b513fb87d802b1167288e1821f60db89359771091770fe8150e5c0c  bitcoin-03a7aaab8409-osx-unsigned.tar.gz
    a5d6b02187573e494b3eb1d613082ba6a225a70725c6bb2d332502ab9ccc9ed2  bitcoin-03a7aaab8409-osx64.tar.gz
    7728c5ba28edf3e69f764f985628faeb85641970cc34460df44e8a8a1fc32e66  src/bitcoin-03a7aaab8409.tar.gz
```

theuni committed:
depends: specify libc++ header location for darwin

For depends builds this was fixed by fbcfcf69, which deleted the conflicting
headers. When we no longer control the clang installation, we need to ensure
that the SDK's libc++ headers are used rather than the ones shipped with clang.
We can do that by turning off the default include path and hard-coding our own.
This hard-coded path is ok because we control (via SDK packaging) where these
headers end-up.

Side-note: Now that this path is hard-coded in depends, we can potentially
package the SDK differently, as the c++ folder can live wherever is most
convenient for us.

theuni commented:
I noticed that this didn't work with vanilla clang 10 from llvm.org. I'm unsure if there's some difference in paths, but vanilla clang picks up the libc++ that it's shipped with over the SDK's libc++. I suspect that distro-shipped clang's are patching default paths, which could explain why it works on some systems and not others.

I was able to fix that like this: https://github.com/theuni/bitcoin/commit/125559aac35f6ad8f6852710d3014379f8d99c4f

@dongcarl I think this is the correct fix, and I suspect it might also simplify #17920 for you. Mind having a look and pulling it in if you agree?

dongcarl committed:
depends: Add justifications for macOS clang flags

dongcarl commented:
Pushed with theuni's patch and an additional commit explaining our long list of macOS clang options. Although this PR worked before theuni's patch, the inclusion of this patch makes our `depends` system more robust on more systems and custom builds of clang, which is great! :relaxed: 

Running Gitian build now...

theuni commented on depends/Makefile:
```diff
@@ -125,6 +125,10 @@ $(host_arch)_$(host_os)_id_string+=$(shell $(host_CXX) --version 2>/dev/null)
 $(host_arch)_$(host_os)_id_string+=$(shell $(host_STRIP) --version 2>/dev/null)
 
+ifneq ($(strip $(FORCE_USE_SYSTEM_CLANG)),)
+$(host_arch)_$(host_os)_id_string+=system_clang
```
Sorry, I realize this was my commit..

In testing, I noticed that we need to add this to the build string too. Because for darwin builds it uses the weird trick of compiling the native builds with the downloaded compiler, we need to redo those builds when we go back and forth between system-clang and non.

```diff
diff --git a/depends/Makefile b/depends/Makefile
index b1d929d8dd..2bc5df974a 100644
--- a/depends/Makefile
+++ b/depends/Makefile
@@ -126,6 +126,7 @@ $(host_arch)_$(host_os)_id_string+=$(shell $(host_RANLIB) --version 2>/dev/null)
 $(host_arch)_$(host_os)_id_string+=$(shell $(host_STRIP) --version 2>/dev/null)

 ifneq ($(strip $(FORCE_USE_SYSTEM_CLANG)),)
+build_id_string+=system_clang
 $(host_arch)_$(host_os)_id_string+=system_clang
 endif
```

dongcarl replied:
Added to: 156b604203ef17b2b77ee9dacf15e375c809242a

theuni commented on depends/Makefile:
```diff
@@ -156,7 +156,11 @@ $(host_arch)_$(host_os)_native_toolchain?=$($(host_os)_native_toolchain)
 include funcs.mk
 
 binutils_path=$($($(host_arch)_$(host_os)_native_binutils)_prefixbin)
+ifeq ($(strip $(FORCE_USE_SYSTEM_CLANG)),)
```
Please add FORCE_USE_SYSTEM_CLANG to the options in README.md (preferably with an expert disclaimer).

dongcarl replied:
See: de4fedb6c3a45e6a23f1bccd28045d76b5830afc

dongcarl committed:
depends: Reformat make options as definition list

dongcarl committed:
depends: Add documentation for FORCE_USE_SYSTEM_CLANG make flag

theuni reviewed (approved):
ACK de4fedb6c3a45e6a23f1bccd28045d76b5830afc.

I've been using this branch for the last week or so, using both external and pinned clang, and all has worked great.

hebasto commented:
@dongcarl 

Multiple warnings observed on Travis macOS 10.12 [build](https://travis-ci.org/github/bitcoin/bitcoin/jobs/718179790#L3561):
`clang: warning: argument unused during compilation: '-nostdinc++' [-Wunused-command-line-argument]`

dongcarl commented:
@fanquake You mentioned on IRC that [hebasto's observation](https://github.com/bitcoin/bitcoin/pull/17919#issuecomment-674404780) was known about. I'm not 100% sure I know what you're talking about though, is there an open issue?

fanquake commented:
> I'm not 100% sure I know what you're talking about though, is there an open issue?

There's no issue, but it was discussed offline IIRC. I'll mention it in #19683, as I think documenting it as part of those changes is probably the easiest thing to do right now.
